Added tests for rate limiting subsets of HTTPRouteRules

martinhesko opened the pull request:
Added tests covering Examples 3 and 4 from the [RateLimitPolicy API v2 proposal](https://docs.kuadrant.io/architecture/rfcs/0001-rlp-v2/). 
As of currently the `test_route_rule_subset` doesn't behave as expected, leaving this as a draft pending further discussion.

UPDATE: apparently the limit for a subset of the HTTPRouteRule is supposed to limit the entire Route Rule if it finds a match.
Added a new test to a separate `method` package to make sure the RLP doesn't apply to HTTPRouteRule with same path but different method

pehala commented on testsuite/tests/kuadrant/limitador/method/test_route_different_method.py:
```diff
@@ -0,0 +1,41 @@
+"""Tests that RLP for a HTTPRouteRule doesn't affect the HTTPRoute with same path but different method"""
```
I do not understand this package, I see no reason why have it in a separate package, there is no fixture that has package scope

martinhesko replied:
We discussed wanting more tests for specific methods with @jsmolar, we can discuss the need for it when he gets back and have this on hold meanwhile.

pehala commented on testsuite/tests/kuadrant/limitador/test_route_rule_subset.py:
```diff
@@ -0,0 +1,38 @@
+
+
+def test_route_rule_subset(client):
+    """Tests that RLP for a subset of HTTPRouteRule applies to the entire HTTPRouteRule"""
```
Is that even correct? this feels wrong

martinhesko replied:
I had a talk about this with @jsmolar because my first assumption was the exact opposite but looking through the ambiguous wording of [Example 4](https://docs.kuadrant.io/architecture/rfcs/0001-rlp-v2/#example-4-targeting-a-subset-of-an-httprouterule-httproutematch-found) we came to the conclusion that it's supposed to work like that

jsmolar replied:
If I remember correctly, we discussed that this is not needed test case and we somehow transformed it into `test_route_different_method.py`, which is correct.

pehala commented on testsuite/tests/kuadrant/limitador/test_route_rule_child.py:
```diff
@@ -0,0 +1,36 @@
+def rate_limit(rate_limit):
+    """Add limit to the policy"""
+    selector = RouteSelector(
+        RouteMatch(path=PathMatch(value="/anything/get", type=MatchType.EXACT), method=HTTPMethod.GET)
```
This is not a child but a subset, it is the more concrete rule. Does it in this case limit if /anything/get is used? If yes, we should test that it correctly limits the subset, if no, I do not see how it is different from any other invalid reference to a rule

martinhesko replied:
It doesn't limit `anything/get`, but what makes it different from an invalid refence would be the `/anything` PathPrefix route rule covering the subset we want to bind our rate policy to despite there not being a specific rule to bind to.

pehala replied:
So I set up rule to "/anything/get" and it does not limit "/anything/get"?

martinhesko replied:
yes, in this configuration `/anything/get` should not be limited because there isn't an exact matching route rule for the RLP, only a route rule matching the paths with prefix `/anything`

martinhesko commented:
After a further discussion ended up deleting one of the cases, renamed the rest and added parametrization to the `test_route_rule_invalid` to test multiple RLP's that shouldn't get applied since they don't have an exact matching HTTPRouteRule
